### PR TITLE
Support transforms for paddle tensor image

### PR DESCRIPTION
**Pull request by lyuwenyu**
### PR types

New features
### PR changes

Others
### Describe

- tensor transforms on GPU
  - normalize
  - to_grayscale
  - vflip
  - hflip
  - crop
  - center_crop
  - pad
  - rotate
  - resize

### CODE REVIEW

**Review by qingqing01 — COMMENTED** [review]
Add unit testing

File: python/paddle/vision/transforms/functional_tensor.py [hunk 1]
```diff
@@ -14,11 +14,95 @@
...
+def _assert_image_tensor(img, data_format):
+    if not isinstance(
+            img, paddle.Tensor) or img.ndim != 3 or not data_format.lower() in (
+                'chw', 'hwc'):
+        raise RuntimeError(
+            'not support [type={}, ndim={}, data_format={}] paddle image'.
+            format(type(img), img.ndim, data_format))
+
+
+# def _assert_data_format(data_format):
```
Comment (heavengate) [inline python/paddle/vision/transforms/functional_tensor.py]: remove

File: python/paddle/vision/transforms/functional_tensor.py [hunk 2]
```diff
@@ -31,10 +115,428 @@ def normalize(img, mean, std, data_format='CHW'):
...
+    img = (img * rgb_weights).sum(axis=_c_index, keepdim=True)
+    _shape = img.shape
+    _shape[_c_index] = num_output_channels
+
+    return img.expand(_shape)
+
+
+def _affine_grid(theta, w, h, ow, oh):
+    d = 0.5
+    # tic = time.time()
```
Comment (heavengate) [inline python/paddle/vision/transforms/functional_tensor.py]: remove comment code

File: python/paddle/vision/transforms/functional_tensor.py [hunk 3]
```diff
@@ -14,11 +14,95 @@
...
+import paddle.nn.functional as F
+
+from paddle.nn.functional import affine_grid, grid_sample
+from paddle.nn.functional import pad as paddle_pad
+
+import sys
+import collections
+
+
+def _assert_image_tensor(img, data_format):
```
Comment (heavengate) [inline python/paddle/vision/transforms/functional_tensor.py:30]: 辅助函数比较多，文件比较长，建议把辅助函数抽出去作为一个`xxx_utils.py`

File: python/paddle/tests/test_transforms.py [hunk 4]
```diff
@@ -285,9 +304,94 @@ def test_info(self):
         str(transforms.Compose([transforms.Resize((224, 224))]))
 
 
-class TestTransformsPIL(TestTransformsCV2):
+class TestTransformsTensor(TestTransformsCV2):
```
Comment (qingqing01) [inline python/paddle/tests/test_transforms.py:312]: Need to keep TestTransformsPIL

File: python/paddle/vision/transforms/functional_tensor.py [hunk 5]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline python/paddle/vision/transforms/functional_tensor.py]: do not change

File: python/paddle/vision/transforms/functional_tensor.py [hunk 6]
```diff
@@ -31,10 +107,426 @@ def normalize(img, mean, std, data_format='CHW'):
...
+    w, h = _get_image_size(img, data_format=data_format)
+
+    img = img if data_format.lower() == 'chw' else img.transpose((0, 3, 1, 2))
+
+    # image center is (0, 0) in matrix
+    # translate = None
+    # if translate is None:
+    #     post_trans = [0, 0]
+    # else:
+    #     post_trans = translate
```
Comment (qingqing01) [inline python/paddle/vision/transforms/functional_tensor.py]: remove these code?

File: python/paddle/vision/transforms/functional_tensor.py [hunk 7]
```diff
@@ -31,10 +107,426 @@ def normalize(img, mean, std, data_format='CHW'):
...
+
+    padding = [pad_left, pad_right, pad_top, pad_bottom]
+
+    if padding_mode == 'edge':
+        padding_mode = 'replicate'
+    elif padding_mode == 'symmetric':
+        raise ValueError('Do not support symmetric mdoe')
+
+    img = img.unsqueeze(0)
+    #  'constant', 'reflect', 'replicate', 'circular'
```
Comment (qingqing01) [inline python/paddle/vision/transforms/functional_tensor.py:456]: 和上面注释的模式 constant, edge, reflect or symmetric 不同？

File: python/paddle/tests/test_transforms.py (hunk 4 above)
Comment (lyuwenyu) [reply python/paddle/tests/test_transforms.py:312]: copy

File: python/paddle/vision/transforms/functional_tensor.py (hunk 5 above)
Comment (lyuwenyu) [reply python/paddle/vision/transforms/functional_tensor.py]: ok

File: python/paddle/vision/transforms/functional_tensor.py (hunk 6 above)
Comment (lyuwenyu) [reply python/paddle/vision/transforms/functional_tensor.py]: ok

File: python/paddle/vision/transforms/functional_tensor.py (hunk 7 above)
Comment (lyuwenyu) [reply python/paddle/vision/transforms/functional_tensor.py:456]: 1. `# 'constant', 'reflect', 'replicate', 'circular'` 这行注释只是为了显示paddle.pad支持的列表，
2. 上边注释里写的`constant, edge, reflect or symmetric `是为了和pil cv的接口保持一致，每种方式都做了处理
3. 也可以支持完整的paddle pad 的 mode  `'constant', 'reflect', 'replicate' 和`'circular'`, **需要改成这种方式嘛？**

**Review by qingqing01 — APPROVED** [review]

**Review by qingqing01 — DISMISSED** [review]

**Review by LielinJiang — APPROVED** [review]
